dhis-2/dhis-e2e-test/src/test/java/org/hisp/dhis/tracker/importer/TrackerExportTests.java

```diff
@@ -126,6 +130,88 @@ public void shouldReturnRequestedFields( String endpoint, String fields, String
             } );
     }
 
+    private Stream<Arguments> shouldReturnRequestedTeiEnrollmentsFields()
+    {
+        return Stream.of( Arguments.of( "enrollments[events, relationships, attributes]", not( emptyIterable() ),
```

Comment:
Do you think we need to write an e2e test for every permutation of these `fields` query parameter values? I think this might add more cost than benefit in terms of security.

You have a unit test for these values translating into the correct object like `TrackerTrackedEntityFieldsParamMapper` that is passed into the querying code.

We have the MockMVC/h2 tests like dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/TrackerTrackedEntitiesExportControllerTest.java which also test the request is met with a response which has these fields set to something non null/empty. This one tests the code querying the DB and using the information we extract out of `fields`.

IMHO these should be enough to ensure we do return these nested fields.

https://github.com/dhis2/dhis2-core/pull/12234 we can now also write controller tests using Postgres. In case you need to test /tracker/enrollments without providing a specific `uid`. As these requests usually rely on JSONB which does not work in the convenience controller tests relying on h2 DB.

Reply:
i don't see any problem with e2e tests, also I've added some of the tests you are mentioning